utils/schemed-kafka-components-utils/src/main/scala/pl/touk/nussknacker/engine/schemedkafka/sink/UniversalKafkaSinkFactory.scala

```diff
@@ -97,146 +103,118 @@ class UniversalKafkaSinkFactory(
+        schemaData = schemaData,
+        validationMode = ValidationMode.lax,
       )
-    case TransformationStep(
-          (`topicParamName`, DefinedEagerParameter(topic: String, _)) ::
-          (`schemaVersionParamName`, DefinedEagerParameter(version: String, _)) ::
-          (`sinkKeyParamName`, _) ::
-          (`sinkRawEditorParamName`, _) ::
-          (`sinkValidationModeParamName`, DefinedEagerParameter(mode: String, _)) ::
-          (`sinkValueParamName`, value: BaseDefinedParameter) :: Nil,
-          _
-        ) =>
-      // edge case - for some reason Topic/Version is not defined
-      getSchema(topic, version)
-        .andThen { runtimeSchemaData =>
-          schemaBasedMessagesSerdeProvider.schemaValidator
-            .validateSchema(runtimeSchemaData.schema)
-            .map(_ => runtimeSchemaData)
-            .leftMap(_.map(e => CustomNodeError(nodeId.id, e.getMessage, None)))
-        }
-        .andThen { runtimeSchemaData =>
-          schemaSupportDispatcher
-            .forSchemaType(runtimeSchemaData.schema.schemaType())
-            .extractParameterForSink(
-              runtimeSchemaData.schema,
-              rawMode = true,
-              validationMode = extractValidationMode(mode),
-              rawParameter = rawValueParamDeclaration.createParameter(),
-              restrictedParamNames
-            )
-            .map { extractedSinkParameter =>
-              val validationAgainstSchemaErrors = extractedSinkParameter
-                .validateParams(Map(sinkValueParamName -> value))
-                .swap
-                .map(_.toList)
-                .getOrElse(List.empty)
-              FinalResults(
-                context,
-                validationAgainstSchemaErrors,
-                Some(TransformationState(runtimeSchemaData, extractedSinkParameter))
-              )
-            }
+        .map { schemaBasedParameter =>
+          val state = TransformationState(schemaData, schemaBasedParameter)
+          NextParameters(schemaBasedParameter.value :: Nil, state = Some(state))
         }
         .valueOr { errors =>
           FinalResults(context, errors.toList)
         }
     case TransformationStep(
           (`topicParamName`, DefinedEagerParameter(_: String, _)) ::
-          (`contentTypeParamName`, DefinedEagerParameter(contentType: String, _)) ::
+          (`contentTypeParamName`, DefinedEagerParameter(_: String, _)) ::
           (`sinkKeyParamName`, _) ::
-          (`sinkRawEditorParamName`, _) ::
-          (`sinkValidationModeParamName`, DefinedEagerParameter(mode: String, _)) ::
           (`sinkValueParamName`, value: BaseDefinedParameter) :: Nil,
+          Some(state)
+        ) =>
+      val validationAgainstSchemaErrors = state.validateParams(Map(sinkValueParamName -> value))
+      FinalResults(
+        context,
+        validationAgainstSchemaErrors,
+        Some(state)
+      )
+  }
+
+  protected def handleTopicWithSchemaWithJsonTemplateEditor(
```

Comment:
Are you still getting what's going on in this class? Because I'm not :) What does `WithJsonTemplateEditor` mean in this context? Does this method handle configuration option when there is a single parameter for value? I see that this pattern matching is enabled even if this option is not turned on. WDYT about having two `PartialFunction` for both options and using one of them depending on configuration option? Will it decompose the problem? Or maybe at least, we could extract some parts of this pattern matching to dedicated classes that would name the scope of this pattern patching that they solve?